rest-api/dao/questionnaire_response_dao.py

```diff
@@ -83,15 +83,51 @@ def _validate_model(self, session, obj):  # pylint: disable=unused-argument
       logging.error(
           'QuestionnaireResponse model has no answers. This is harmless but probably an error.')
 
+  def _validate_link_ids_from_resource_json_group(self, resource, link_ids):
+    """
+    Look for question sections and validate the linkid in each answer. If there is a response
+    answer link id that does not exist in the questionnaire, then raise an exception.
+    This is a recursive function because answer groups can be nested.
+    :param resource: A group section of the response json.
+    :param link_ids: List of link ids to validate against.
+    """
+    # look for child group sections and call ourselves
+    if 'group' in resource:
+      self._validate_link_ids_from_resource_json_group(resource['group'], link_ids)
+
+    if type(resource) is list:
+      for item in resource:
+        self._validate_link_ids_from_resource_json_group(item, link_ids)
+
+    # look for question section:
+    if 'question' in resource:
+
+      # if question is a list, then loop through the list
+      if type(resource['question']) is list:
+        self._validate_link_ids_from_resource_json_group(resource['question'], link_ids)
+
+      # if question is not a list, check linkId directly
+      for question in resource['question']:
+        # Do not raise exception when link id is 'ignoreThis' for unit testing.
+        if 'linkId' in question and question['linkId'].lower() != 'ignorethis' \
```

Comment:
optional: test-only sentinel values in prod code should definitely be avoided, though understood that changing this might be a massive refactoring task we don't want to take on currently. Consider whether there's another straight forward way to achieve this.

Reply:
That one check fixed a lot of unit test breakage. The primary issue here is the unit test json data files for questionnaire/responses look (to me) like they were from early iterations of the module.  They have little resemblance to the format of the recent questionnaire/response json data in stable or production.  Refactoring all those data files and the unit tests would be a good sized task, probably suitable for a separate ticket.